pkg/storage/ingest/fetcher.go

```diff
@@ -555,37 +622,42 @@ func sumRecordLengths(records []*kgo.Record) (sum int) {
 func (r *concurrentFetchers) run(ctx context.Context, wants chan fetchWant, logger log.Logger, highWatermark *atomic.Int64) {
 	defer r.wg.Done()
 
-	errBackoff := backoff.New(ctx, backoff.Config{
-		MinBackoff: 250 * time.Millisecond,
-		MaxBackoff: 2 * time.Second,
-		MaxRetries: 0, // retry forever
-	})
+	errBackoff := backoff.New(ctx, r.fetchBackoffConfig)
 
 	for w := range wants {
 		// Start new span for each fetchWant. We want to record the lifecycle of a single record from being fetched to being ingested.
 		wantSpan, ctx := spanlogger.NewWithLogger(ctx, logger, "concurrentFetcher.fetch")
 		wantSpan.SetTag("start_offset", w.startOffset)
 		wantSpan.SetTag("end_offset", w.endOffset)
 
-		var previousResult fetchResult
+		// This current buffered fetchResult that has not been sent to the result channel yet.
+		// This is empty at the beginning, then we merge records as soon as we receive them
+		// from the Fetch response(s).
+		var bufferedResult fetchResult
+
 		for attempt := 0; errBackoff.Ongoing() && w.endOffset > w.startOffset; attempt++ {
 			attemptSpan, ctx := spanlogger.NewWithLogger(ctx, logger, "concurrentFetcher.fetch.attempt")
 			attemptSpan.SetTag("attempt", attempt)
 
-			f := r.fetchSingle(ctx, w)
-
-			// We increase the count of buffered records as soon as we fetch them.
-			r.bufferedFetchedRecords.Add(int64(len(f.Records)))
+			// Run a single Fetch request.
+			if res := r.fetchSingle(ctx, w); res.Err != nil {
+				// We got an error. We handle it and then discard this fetch result content.
+				w = handleKafkaFetchErr(res.Err, w, errBackoff, r.startOffsets, r.client, attemptSpan)
+			} else {
+				// We increase the count of buffered records as soon as we fetch them.
+				r.bufferedFetchedRecords.Add(int64(len(res.Records)))
+
+				// Update the high watermark.
+				if hwm := res.HighWatermark; hwm >= 0 {
+					casHWM(highWatermark, hwm)
+				}
 
-			f = f.Merge(previousResult)
-			previousResult = f
-			if f.Err != nil {
-				w = handleKafkaFetchErr(f.Err, w, errBackoff, r.startOffsets, r.client, attemptSpan)
+				// Merge the last fetch result if the previous buffered result (if any).
+				// Keep non-mergeable fields from res, because the last response is the most updated one.
+				bufferedResult = res.Merge(bufferedResult)
 			}
-			if hwm := f.HighWatermark; hwm >= 0 {
-				casHWM(highWatermark, hwm)
-			}
-			if len(f.Records) == 0 {
+
+			if len(bufferedResult.Records) == 0 {
 				// Typically if we had an error, then there wouldn't be any records.
```

Comment:
this comment is now outdated.

i couldn't find a claim that an error means no records. Are we relying on the alert for offset gaps to validate this assumption?

Reply:
> this comment is now outdated.

Commented updated here: https://github.com/grafana/mimir/pull/9963/commits/883e4ea9aeae2604303991da9a9465491b809da6